File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/NativeblocksFrame.kt
package io.nativeblocks.runtime.api

import androidx.compose.foundation.background
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import io.nativeblocks.runtime.di.NativeCoreSDKInjector
import io.nativeblocks.runtime.frame.FrameViewModel
import io.nativeblocks.runtime.frame.NativeFrame
import org.koin.androidx.compose.koinViewModel
import org.koin.compose.KoinIsolatedContext
import org.koin.core.qualifier.named

/**
 * Composable function for rendering a Nativeblocks frame.
 * Sets up the frame with the provided route and arguments, and handles loading and error states.
 *
 * @param route The route of the frame to display.
 * @param arguments A map of arguments to pass to the frame route.
 * @param state Whether the frame keeps what the user did to it, and under which key.
 * @param loading A composable to display during the loading state.
 * @param error A composable to display in case of an error, accepting an error message.
 */
@Composable
fun NativeblocksFrame(
    instanceName: String = "default",
    route: String,
    arguments: Map<String, String>,
    state: NativeblocksFrameState = NativeblocksFrameState.Stateless,
    loading: @Composable () -> Unit,
    error: @Composable (String) -> Unit
) {
    KoinIsolatedContext(context = NativeCoreSDKInjector.get(instanceName)) {
        val systemContractors = NativeblocksManager.getInstance(instanceName).providedActionContractors()
        systemContractors.forEach {
            it.ActionContractor()
        }
        val frameViewModel = koinViewModel<FrameViewModel>(
            qualifier = named(instanceName),
            key = "$instanceName-$route-${arguments.hashCode()}"
        )
        LaunchedEffect(instanceName, route, arguments, state) {
            frameViewModel.setupFrame(route, arguments, state.key)
        }
        NativeFrame(instanceName, frameViewModel, loading, error)
    }
}

/**
 * Composable function for displaying a loading indicator in Nativeblocks.
 * Adjusts the background and indicator color based on the system's dark mode setting.
 */
@Composable
fun NativeblocksLoading() {
    val bgColor = if (isSystemInDarkTheme()) Color.Gray else Color.White
    Column(
        Modifier
            .fillMaxSize()
            .background(bgColor),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Column {
            val color = if (isSystemInDarkTheme()) Color.White else Color.Black
            CircularProgressIndicator(color = color)
        }
    }
}

/**
 * Composable function for displaying an error message in Nativeblocks.
 * Adjusts the background and text color based on the system's dark mode setting.
 *
 * @param message The error message to display.
 */
@Composable
fun NativeblocksError(message: String) {
    val bgColor = if (isSystemInDarkTheme()) Color.Gray else Color.White
    Column(
        Modifier
            .fillMaxSize()
            .background(bgColor),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        val color = if (isSystemInDarkTheme()) Color.White else Color.Black
        Text(text = "Something went wrong!", color = color)
        Text(text = message, color = color)
    }
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/NativeblocksFrame.swift
import Foundation
import SwiftUI

/// Represents a view that manages and displays a Nativeblocks frame.
/// The `NativeblocksFrame` is used to display a portion of the UI as defined by a specific route and its arguments.
public struct NativeblocksFrame: View {
    private var instanceName: String = "default"
    private var route: String
    private var routeArguments: [String: String]
    private var state: NativeblocksFrameState
    private var loading: () -> AnyView
    private var error: (String) -> AnyView

    /// Initializes a new `NativeblocksFrame` with the specified parameters.
    /// - Parameters:
    ///   - route: The route to be loaded.
    ///   - routeArguments: A dictionary of arguments to pass to the route.
    ///   - state: Whether the frame keeps what the user did to it, and under which key.
    ///   - loading: A closure that returns a view to be shown while the content is loading.
    ///   - error: A closure that takes a message string and returns a view to be shown when an error occurs.
    public init(
        instanceName: String = "default",
        route: String,
        routeArguments: [String: String],
        state: NativeblocksFrameState = .stateless,
        loading: @escaping () -> AnyView,
        error: @escaping (String) -> AnyView
    ) {
        self.instanceName = instanceName
        self.route = route
        self.routeArguments = routeArguments
        self.state = state
        self.loading = loading
        self.error = error
    }

    public var body: some View {
        let contractors = NativeblocksManager.getInstance(name: instanceName).providedActionContractors()
        ZStack {
            ForEach(Array(contractors.enumerated()), id: \.offset) { _, contractor in
                AnyView(contractor.actionContractor())
            }
            NativeFrame(
                instanceName: instanceName,
                route: route,
                args: routeArguments,
                state: state,
                loading: loading,
                error: error
            )
        }
    }
}

/// Represents a view displayed during the loading state of a `NativeblocksFrame`.
public struct NativeblocksLoading: View {
    public init() {}

    public var body: some View {
        #if os(iOS)
            let bgColor = Color(UIColor.systemBackground)
        #elseif os(macOS)
            let bgColor = Color(NSColor.windowBackgroundColor)
        #endif

        VStack {
            ProgressView()
                .progressViewStyle(CircularProgressViewStyle())
                .foregroundColor(Color.primary)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(bgColor)
    }
}

/// Represents a view displayed when an error occurs while loading a `NativeblocksFrame`.
public struct NativeblocksError: View {
    private var message: String

    /// Initializes a new `NativeblocksError` view with the specified message.
    /// - Parameter message: The error message to display.
    public init(message: String) {
        self.message = message
    }

    public var body: some View {
        #if os(iOS)
            let bgColor = Color(UIColor.systemBackground)
        #elseif os(macOS)
            let bgColor = Color(NSColor.windowBackgroundColor)
        #endif

        VStack {
            Text("Something went wrong!")
                .foregroundColor(Color.primary)
            Text(message)
                .foregroundColor(Color.primary)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(bgColor)
    }
}
